src/InteractiveWindow/Editor/History.cs

```diff
@@ -155,82 +159,87 @@ private Entry Get(string pattern, int step)
             }
 
             return next;
-        }
+         }
 
-        internal Entry GetNext(string pattern = null)
+        internal Entry GetPrevious(string pattern)
         {
-            return Get(pattern, step: +1);
-        }
+            var startPos = _current;
+            Entry next;
+            next = MovePrevious(pattern);
+            if (next == null)
+            {
+                _current = startPos;
+                return null;
+            }
 
-        internal Entry GetPrevious(string pattern = null)
-        {
-            return Get(pattern, step: -1);
-        }
+            return next;
+         }
 
-        private Entry Move(string pattern, int step)
+        private Entry MoveNext(string pattern)
         {
-            Debug.Assert(step == -1 || step == +1);
+            if (Length == 0) return null;
+
+            bool wasUninitialized = (_current == -1);
+            if (wasUninitialized) return null;
 
-            bool wasLive = _live;
             _live = true;
 
-            if (Length == 0)
+            int end = Length - 1;
+            for (_current++; _current <= end; _current++)
             {
-                return null;
+                Entry entry;
+                if (TryMatch(pattern, out entry)) return entry;
             }
 
-            bool patternEmpty = string.IsNullOrWhiteSpace(pattern);
+            return null;
+        }
 
-            int start, end;
-            if (step > 0)
+        private Entry MovePrevious(string pattern)
```

Comment:
I think you could inline ```MoveNext``` and ```MovePrevious``` but I'll understand if you prefer to leave that for later.